### PR TITLE
support external-dns on ingress objects

### PR DESCRIPTION
**Pull request by bavarianbidi**
to support dns-record creation via external-dns, two additional annotations are needed to make external-dns work on pmo created ingress objects.

towards: https://github.com/giantswarm/roadmap/issues/1037
related `config` repo change: https://github.com/giantswarm/config/pull/1603

TODO Mario:

- [ ] test this version of PMO again

## Checklist

I have:

- [x] Described why this change is being introduced
- [ ] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

File: helm/prometheus-meta-operator/values.schema.json [hunk 1]
```diff
@@ -232,4 +240,4 @@
             "type": "string"
         }
     }
-}
+}
```
Comment (QuentinBisson) [inline helm/prometheus-meta-operator/values.schema.json:243]:
```suggestion
}

```

File: service/controller/resource/monitoring/ingress/resource.go [hunk 2]
```diff
@@ -60,8 +60,10 @@ func getObjectMeta(v interface{}, config Config) (metav1.ObjectMeta, error) {
 	}
 
 	annotations := map[string]string{
-		"nginx.ingress.kubernetes.io/auth-signin": "https://$host/oauth2/start?rd=$escaped_request_uri",
-		"nginx.ingress.kubernetes.io/auth-url":    "https://$host/oauth2/auth",
+		"nginx.ingress.kubernetes.io/auth-signin":   "https://$host/oauth2/start?rd=$escaped_request_uri",
+		"nginx.ingress.kubernetes.io/auth-url":      "https://$host/oauth2/auth",
+		"external-dns.alpha.kubernetes.io/hostname": config.BaseDomain,
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/ingress/resource.go:65]: Should this not be added only if `.Values.ingress.externalDNS` is true?

File: service/key/key.go [hunk 3]
```diff
@@ -151,6 +151,9 @@ func RemoteWriteAuthenticationAnnotations() map[string]string {
 		"nginx.ingress.kubernetes.io/client-body-buffer-size": "10m",
 		// Remote write requests can be quite big. (default max body size: 1m)
 		"nginx.ingress.kubernetes.io/proxy-body-size": "10m",
+		// external-dns
+		"external-dns.alpha.kubernetes.io/hostname": baseDomain,
```
Comment (QuentinBisson) [inline service/key/key.go:155]: Same remark as above

File: service/controller/resource/monitoring/ingress/resource.go (hunk 2 above)
Comment (bavarianbidi) [reply service/controller/resource/monitoring/ingress/resource.go:65]: yes, for sure - just wanted your confirmation on this :sweat_smile:

**Review by QuentinBisson — APPROVED** [review]
LGTM if tested :)

**Comment by bavarianbidi** [comment]
@QuentinBisson i had to add one additional commit and now it looks very good :sweat_smile: 

I've tested the following:

Same version of `PMO`, once with:
```yaml
ingress:
  externalDNS:
    enabled: false
```
This is the default for all existing installations and you have to opt-in if the external-dns annotations should be created by setting
```yaml
ingress:
  externalDNS:
    enabled: true
```


```
➜ k get ingress -A -o yaml | grep -E "external-dns.alpha.kubernetes.io" | wc -l                                                                                  
14
➜ k get ingress -A -o yaml | grep -E "external-dns.alpha.kubernetes.io" | wc -l                                                                    
10
```

```diff
--- /tmp/pmo/disabled-external-dns/glippy-prometheus_prometheus-glippy.yaml	2023-03-02 10:38:37.321023608 +0100
+++ /tmp/pmo/enabled-external-dns/glippy-prometheus_prometheus-glippy.yaml	2023-03-02 10:30:30.186118861 +0100
@@ -2,6 +2,8 @@
 kind: Ingress
 metadata:
   annotations:
+    external-dns.alpha.kubernetes.io/hostname: prometheus.glippy.azuretest.gigantic.io
+    giantswarm.io/external-dns: managed
     nginx.ingress.kubernetes.io/auth-signin: https://$host/oauth2/start?rd=$escaped_request_uri
     nginx.ingress.kubernetes.io/auth-url: https://$host/oauth2/auth
   creationTimestamp: "2023-02-21T15:46:21Z"
@@ -13,7 +15,7 @@
     giantswarm.io/cluster: glippy
   name: prometheus-glippy
   namespace: glippy-prometheus
-  resourceVersion: "25373475"
+  resourceVersion: "25013586"
   uid: 5dab1140-65a8-4b1e-b1d0-08ae97cfcebe
 spec:
   ingressClassName: nginx
```

```diff
--- /tmp/pmo/disabled-external-dns/glippy-prometheus_prometheus-glippy-remote-write.yaml	2023-03-02 10:38:19.969074563 +0100
+++ /tmp/pmo/enabled-external-dns/glippy-prometheus_prometheus-glippy-remote-write.yaml	2023-03-02 10:30:50.182092208 +0100
@@ -2,6 +2,8 @@
 kind: Ingress
 metadata:
   annotations:
+    external-dns.alpha.kubernetes.io/hostname: prometheus.glippy.azuretest.gigantic.io
+    giantswarm.io/external-dns: managed
     nginx.ingress.kubernetes.io/auth-realm: Authentication Required
     nginx.ingress.kubernetes.io/auth-secret: remote-write-ingress-auth
     nginx.ingress.kubernetes.io/auth-type: basic
@@ -16,7 +18,7 @@
     giantswarm.io/cluster: glippy
   name: prometheus-glippy-remote-write
   namespace: glippy-prometheus
-  resourceVersion: "25373418"
+  resourceVersion: "25013530"
   uid: ab658886-c5ea-4499-bb5c-fa114002c5e6
 spec:
   ingressClassName: nginx
```

```diff
--- /tmp/pmo/disabled-external-dns/monitoring_prometheus-meta-operator.yaml	2023-03-02 10:38:09.277105641 +0100
+++ /tmp/pmo/enabled-external-dns/monitoring_prometheus-meta-operator.yaml	2023-03-02 10:31:24.002042500 +0100
@@ -2,6 +2,8 @@
 kind: Ingress
 metadata:
   annotations:
+    external-dns.alpha.kubernetes.io/hostname: alertmanager.glippy.azuretest.gigantic.io,prometheus.glippy.azuretest.gigantic.io
+    giantswarm.io/external-dns: managed
     kubernetes.io/tls-acme: "true"
     meta.helm.sh/release-name: prometheus-meta-operator
     meta.helm.sh/release-namespace: monitoring
@@ -17,7 +19,7 @@
     helm.sh/chart: prometheus-meta-operator-4.23.0-8bf8b2b1a8fb29fb8643612f77dccc2
   name: prometheus-meta-operator
   namespace: monitoring
-  resourceVersion: "25371546"
+  resourceVersion: "25100021"
   uid: 3fd90ce3-faa6-4dc5-b7f0-bc48cb015fbb
 spec:
   ingressClassName: nginx
```

```diff
--- /tmp/pmo/disabled-external-dns/monitoring_prometheus-meta-operator-alertmanager.yaml	2023-03-02 10:37:59.053135132 +0100
+++ /tmp/pmo/enabled-external-dns/monitoring_prometheus-meta-operator-alertmanager.yaml	2023-03-02 10:31:34.550025860 +0100
@@ -2,6 +2,8 @@
 kind: Ingress
 metadata:
   annotations:
+    external-dns.alpha.kubernetes.io/hostname: alertmanager.glippy.azuretest.gigantic.io
+    giantswarm.io/external-dns: managed
     kubernetes.io/tls-acme: "true"
     meta.helm.sh/release-name: prometheus-meta-operator
     meta.helm.sh/release-namespace: monitoring
@@ -19,7 +21,7 @@
     helm.sh/chart: prometheus-meta-operator-4.23.0-8bf8b2b1a8fb29fb8643612f77dccc2
   name: prometheus-meta-operator-alertmanager
   namespace: monitoring
-  resourceVersion: "25371531"
+  resourceVersion: "25099977"
   uid: dc26ccdb-4656-4919-aac4-5defd32b6a78
 spec:
   ingressClassName: nginx
```

**Review by QuentinBisson — APPROVED** [review]

**Pull request merged**